crates/ty_ide/src/docstring.rs

```diff
@@ -153,6 +149,128 @@ fn documentation_trim(docs: &str) -> String {
     output
 }
 
+/// Given a cleaned up docstring, render it to markdown.
+///
+/// This is by no means perfect but it handles a lot of common formats well.
+///
+/// This code currently assumes docstrings are reStructuredText with significant
+/// line-breaks and indentation. Some constructs like *italic*, **bold**, and `inline code`
+/// are shared between the two formats. Other things like _italic_ are only markdown
+/// and should be escaped.
+fn render_markdown(docstring: &str) -> String {
+    // TODO: there is a convention that `singletick` is for items that can
+    // be looked up in-scope while ``multitick`` is for opaque inline code.
+    // While rendering this we should make note of all the `singletick` locations
+    // and (possibly in a higher up piece of logic) try to resolve the names for
+    // cross-linking. (Similar to `TypeDetails` in the type formatting code.)
+    let mut output = String::new();
+    let mut first_line = true;
+    let mut block_indent = 0;
+    let mut in_doctest = false;
+    let mut starting_literal = false;
+    let mut in_literal = false;
+    let mut in_any_code = false;
+    for untrimmed_line in docstring.lines() {
```

Comment:
This is the worst parser I've written in a long while so if you tell me "Aria for the love of god make a Parser type with methods and enums" I will. Writing it this way was conducive to experimentation and thinking about the format.